tests/Replication2/ReplicatedState/ReplicatedStateTest.cpp

```diff
@@ -0,0 +1,262 @@
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Lars Maier
+////////////////////////////////////////////////////////////////////////////////
+
+#include <gtest/gtest.h>
+
+#include "Mocks/LogLevels.h"
+#include "Replication2/ReplicatedLog/TestHelper.h"
+
+#include "Replication2/Streams/Streams.h"
+#include "Replication2/Streams/LogMultiplexer.h"
+#include "Replication2/ReplicatedState/ReplicatedStateFeature.h"
+#include "Replication2/ReplicatedState/ReplicatedState.h"
+
+#include "StateMachines/MyStateMachine.h"
+
+using namespace arangodb;
+using namespace arangodb::replication2;
+using namespace arangodb::replication2::replicated_state;
+using namespace arangodb::replication2::test;
+
+
+struct ReplicatedStateTest
+    : test::ReplicatedLogTest{
+  ReplicatedStateTest() {
+    feature->registerStateType<MyState>("my-state");
+  }
+  std::shared_ptr<ReplicatedStateFeature> feature =
+      std::make_shared<ReplicatedStateFeature>();
+};
+
+TEST_F(ReplicatedStateTest, simple_become_follower_test) {
+  auto log = makeReplicatedLog(LogId{1});
+  auto follower = log->becomeFollower("follower", LogTerm{1}, "leader");
+  auto state = std::dynamic_pointer_cast<ReplicatedState<MyState>>(
+      feature->createReplicatedState("my-state", log));
+  ASSERT_NE(state, nullptr);
+
+  state->flush();
+
+  auto leaderLog = makeReplicatedLog(LogId{1});
+  auto leader = leaderLog->becomeLeader(LogConfig(2, 2, 2, false), "leader",
+                                        LogTerm{1}, {follower});
+  auto mux = streams::LogMultiplexer<replicated_state::ReplicatedStateStreamSpec<MyState>>::construct(leader);
+  auto inputStream = mux->getStreamById<1>();
+
+  inputStream->insert({.key = "hello", .value = "world"});
+  while (follower->hasPendingAppendEntries()) {
+    follower->runAsyncAppendEntries();
+  }
+
+  auto followerState = state->getFollower();
+  ASSERT_NE(followerState, nullptr);
+  auto& store = followerState->store;
+  EXPECT_EQ(store.size(), 1);
+  EXPECT_EQ(store["hello"], "world");
+}
+
+TEST_F(ReplicatedStateTest, recreate_follower_on_new_term) {
+  auto log = makeReplicatedLog(LogId{1});
+  auto follower = log->becomeFollower("follower", LogTerm{1}, "leader");
+  auto state = std::dynamic_pointer_cast<ReplicatedState<MyState>>(
+      feature->createReplicatedState("my-state", log));
+  ASSERT_NE(state, nullptr);
+
+  // create a leader in term 1
+  auto leaderLog = makeReplicatedLog(LogId{1});
+  auto leader = leaderLog->becomeLeader(LogConfig(2, 2, 2, false), "leader",
+                                        LogTerm{1}, {follower});
+  auto mux = streams::LogMultiplexer<ReplicatedStateStreamSpec<MyState>>::construct(leader);
+  auto inputStream = mux->getStreamById<1>();
+  inputStream->insert({.key = "hello", .value = "world"});
```

Comment:
This looks like it wouldn't affect the test in any way. Just pointing it out, not sure if it's left intentionally.

Reply:
oh it does. We assert that this entry was applied to the state in line 110 to 111.